### ISSUE TITLE
[ttnn/data_movement] slice returns rows from one input row too early when a non-innermost dim is truncated and the innermost begin is misaligned

### ISSUE THREAD

**Issue by jasonmacTT** [tenstorrent/tt-metal#57382]
**Component / Area:** ttnn data_movement — `slice` (row-major program factory)

**Issue Type:** Bad Outputs (silent wrong values)

## Observed

`ttnn.slice` silently returns wrong data. No error is raised and the output shape is correct.

```python
import torch, ttnn
d = ttnn.open_device(device_id=0)

SH = (5, 65, 1025)
x = torch.arange(SH[0]*SH[1]*SH[2], dtype=torch.float32).reshape(SH)
a = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.ROW_MAJOR_LAYOUT, device=d)

got  = ttnn.to_torch(ttnn.slice(a, [0, 0, 1], [5, 64, 1025], [1, 1, 1]))
want = x[0:5, 0:64, 1:1025]
print(int((got != want).sum()))   # 6144, expected 0

ttnn.close_device(d)
```

The input has **65 rows per batch** and the output **64**, so one input row per batch must be
skipped. The failure is that the skip is missed at some batch boundaries: the op returns the row
it was supposed to discard.

6144 elements wrong (6 of 320 output rows). Every bad row contains the input row **immediately
before** the correct one — a linear offset of exactly `-1025`, one input row stride:

```
out[1,0] contains input[0,64]    expected input[1,0]
out[2,0] contains input[1,64]    expected input[2,0]
out[2,1] contains input[2,0]     expected input[2,1]
out[3,0] contains input[2,64]    expected input[3,0]
out[3,1] contains input[3,0]     expected input[3,1]
out[3,2] contains input[3,1]     expected input[3,2]
```

Batches 0 and 4 are entirely correct. Each affected batch is wrong for only its first few rows,
then resyncs.

With `arange` data, where `input[b,r,w] = b*66625 + r*1025 + w`:

```
out[1,0] should be   66626 66627 66628 ...   (input[1,0],  from column 1)
out[1,0] actually is 65601 65602 65603 ...   (input[0,64], from column 1)
```

Identical results for `ROW_MAJOR_LAYOUT` and `TILE_LAYOUT`.

With `bfloat16` instead of `float32`, 1024 elements (1 row, `out[1,0]`) are wrong — the corrupted
row **count** is dtype-dependent.

## Bisect

| tt-metal commit | f32 | bf16 |
| --- | --- | --- |
| `59b5c26c63c8a60011486a1124cdd431dbc200bf` (= `70ba57dc3ab^`) | 0 bad | 0 bad |
| `cffd17520cdfc45dc85c1c47cc2aa49f6bff15c0` (contains `70ba57dc3ab`) | 6144 bad | 1024 bad |

`70ba57dc3ab5d8acaeef2fe059646eea423bcb0f` is the Metal 2.0 port of `data_movement/slice`
(https://github.com/tenstorrent/tt-metal/pull/56494), labelled **"No functional or performance
change intended."** It is the only commit in that range touching
`ttnn/cpp/ttnn/operations/data_movement/slice/`, so it is the prime suspect.

## Trigger conditions

Both are required — either one alone produces correct output:

| variation on `(5,65,1025)` f32 | bad elements |
| --- | --- |
| innermost begin = 1, middle dim truncated 65→64 | **6144** |
| innermost begin = **0**, middle dim truncated 65→64 | 0 |
| innermost begin = 1, middle dim **not** truncated | 0 |
| innermost begin = 2, middle dim truncated | 6138 |

The dependence on a nonzero innermost begin points at the misaligned read path
(`misalignment = begins_bytes % src_buffer_alignment` in `slice_program_factory_rm.cpp`).

## Why only some rows — work-split dependent

Corruption appears exactly when a core's assigned stick range straddles a batch boundary:

| output rows | rows/core (÷64) | divides 64 rows-per-batch? | bad rows |
| --- | --- | --- | --- |
| 320 (5 batches) | 5 | no | 6 |
| 192 (3 batches) | 3 | no | 1 |
| 256 (4 batches) | 4 | yes | **0** |
| 128 (2 batches) | 2 | yes | **0** |
| 64 (1 batch) | 1 | yes | **0** |

When rows-per-core divides the per-batch row count, every core starts on a batch boundary and the
output is correct.

A core that starts inside one batch and runs into the next appears to **fail to skip the leftover
input row** at the crossing, so every stick it produces after that point lags by exactly one input
row. The next core computes its start id from scratch and is correct again — which is why each
affected batch is wrong only for its first few rows.

This also explains the dtype dependence: bf16 halves the bytes per stick, changing the read
chunking and therefore how many cores straddle a boundary.

## Impact

Silent wrong data with no error. `bfloat16` is affected too, but corrupts few enough elements to
stay above a 0.99 PCC threshold, so PCC-gated tests **pass while returning bad data**. This is why
it went unnoticed — it surfaced only because one f32 case corrupted enough rows to drop PCC to
0.9814.

## Where this was hit

Found while uplifting the tt-metal pin in tt-mlir to `cffd17520cdfc45dc85c1c47cc2aa49f6bff15c0`:

- uplift PR: https://github.com/tenstorrent/tt-mlir/pull/9354
- failing test: `test/python/golden/ttir_ops/data_movement/test_data_movement.py::test_slice`,
  parameter `((5, 65, 1025), [0, 0, 1], [5, 64, 1025], None)` with `dtype=float32`
  (`actual_pcc = 0.9814 < 0.99`)

## Environment

- **Hardware:** Wormhole n150 (single card)
- **tt-metal:** `cffd17520cdfc45dc85c1c47cc2aa49f6bff15c0`, built from source
- Both the failing and the clean build were run on the same machine in the same session

## Not established

- Not tested on Blackhole / p150.
- Only `float32` and `bfloat16` tested; other dtypes not checked.

**Comment by akshat-lakhera** [comment]
I'm looking into this row-major stride offset bug in slice_op.cpp and setting up a reproducer test. Will put up a draft PR once verified.
